### ISSUE TITLE
[$3500] Upgrade to react-native-web v0.19.6 or greater

### ISSUE THREAD

**Issue by roryabraham** [Expensify/App#16660]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

### Problem

There are a number of initiatives and bug fixes that depend on React Native 0.19. We can aggregate those all here:

- https://github.com/Expensify/App/issues/11680
- https://github.com/Expensify/App/issues/14139
- https://github.com/Expensify/App/issues/14848

### Solution

Upgrade E/App to use react-native-web 0.19.

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01a99afc2bafb0eada</li>
        <li>Upwork Job ID: 1668778453001940992</li>
        <li>Last Price Increase: 2023-12-01</li>
<li>Automatic offers: </li>
<ul>
<li>aimane-chnaif | Reviewer | 26072994</li>
<li>getusha | Contributor | 26072996</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @kevinksullivan</details>

**Comment by roryabraham** [comment]
I [synced our fork with the upstream](https://github.com/Expensify/react-native-web/pull/17) yesterday, so our fork officially contains the 0.19 code. It's not published yet though. Assigning to myself to publish our react-native-web fork when this comes off HOLD.

**Comment by roryabraham** [comment]
We had a couple of VirtualizedList fixes from https://github.com/Expensify/App/issues/12054 included in 19.2 🎉

**Comment by roryabraham** [comment]
Still on HOLD

**Comment by roryabraham** [comment]
Still on HOLD

**Comment by cristipaval** [comment]
FYI: I added [one more issue](https://github.com/Expensify/App/issues/14848) in the issue description, in the list of the issues held on this one.

**Comment by roryabraham** [comment]
No update here, still on HOLD for the [Accessible Pressables](https://github.com/orgs/Expensify/projects/29/views/1?layout=board) project, which is making steady progress. Technically we don't need to complete the project, we just need to get to a point where there's no more `Touchable` components.

**Comment by janicduplessis** [comment]
I opened a [PR to sync again](https://github.com/Expensify/react-native-web/pull/19), this will unblock https://github.com/Expensify/App/pull/19345 as one of the new release fixes an issue with batchSize > 1 and inverted FlatList.

@roryabraham What is the reason for the hold on the pressables project?

**Comment by janicduplessis** [comment]
I just verified and they got deprecated and not removed so it could still be usable before the Touchable migration project is done. We could add the warning to logbox ignore list until the migration is over.

**Comment by roryabraham** [comment]
At this point, there are only 5 remaining `Touchable` components across the app, so let's just finish that migration then take this off hold early next week

**Comment by roryabraham** [comment]
https://github.com/Expensify/App/issues/17019 is the last remaining PR that should be merged before this comes off HOLD.

**Comment by roryabraham** [comment]
aaaand that's been merged. So we are good to push this one forward now.

**Comment by roryabraham** [comment]
Might as well make this one external now for the actual upgrade, eh?

**Comment by syedsaroshfarrukhdot** [comment]
@roryabraham Can you assign me this issue.

**Comment by roryabraham** [comment]
@syedsaroshfarrukhdot This one is open for proposals, but please review our contributing guidelines and post a proposal using our proposal template.

**Comment by roryabraham** [comment]
Also keep in mind that there are many considerations here, please thoroughly review the deprecations and breaking changes in the release notes, and include a detailed analysis of which ones affect our app, how we can address them, and when we need to.

**Comment by tungmt** [comment]
## Proposal

###  Please re-state the problem that we are trying to solve in this issue.
Upgrade E/App to use react-native-web 0.19.

### What is the root cause of that problem?
There are a number of initiatives and bug fixes that depend on react-native-web 0.19.x

### What changes do you think we should make in order to solve the problem?
There is some functions/components deprecated need to be replaced

- TextInput prop numberOfLines. Use rows prop.
- TextInput prop editable. Use readOnly prop.
- TextInput prop returnKeyType. Use enterKeyHint prop.
- TextInput prop keyboardType. Use inputMode prop.
- Deprecate textAlignVertical style. Use verticalAlign style.
- Deprecate selectable prop. Use userSelect style.
- Deprecate pointerEvents prop. Use pointerEvents style.
- Deprecate focusable props. Use tabIndex prop.
- Deprecate nativeId prop. Use id prop.
- Deprecate accessibility* props. Use aria-* and role props.
- Deprecate StyleSheet.compose(). Use style array syntax.

### What alternative solutions did you explore? (Optional)

**Comment by roryabraham** [comment]
@tungmt I think I'd like to see a bit more details specific to our app in your proposal. Also a big breaking change that wasn't mentioned in your proposal is that `<Text>` is position:relative by default.

**Comment by tungmt** [comment]
Hi @roryabraham,
I updated my proposal https://github.com/Expensify/App/issues/16660#issuecomment-1591366594

**Comment by mountiny** [comment]
@aimane-chnaif Would be great to get this change out sooner than later as it will unblock some other improvements. Would you be bale to work with @tungmt on the proposal details?

**Comment by tungmt** [comment]
> Please do notice that we have a few deprecations in RN_web which depend on the React-native latest version. So There might be some props that are only present in newer RN versions and upgrading RN-web will force us to use those. If our RN version does not have those props then it might create issues.

As I check all RN_web 0.19 deprecated change is supported by RN 0.71.0 and above. So I think we dont need to force update RN in this case.

**Comment by roryabraham** [comment]
Thanks for the heads-up @parasharrajat. We are upgrading our react-native version in https://github.com/Expensify/App/pull/18507

**Comment by conorpendergrast** [comment]
Looks like @aimane-chnaif is still working on reviewing the proposal here!

**Comment by roryabraham** [comment]
Still looking for an approved proposal here.

**Comment by conorpendergrast** [comment]
@aimane-chnaif How is the proposal review going? What's your plan for getting that accepted/ rejected?

**Comment by conorpendergrast** [comment]
Excellent, thank you for the update!

**Comment by conorpendergrast** [comment]
We're still waiting on an approved solution here, through testing by @aimane-chnaif !

**Comment by roryabraham** [comment]
@aimane-chnaif is also assigned to review https://github.com/Expensify/App/pull/18507, which is a pre-requisite for this and therefore takes precedence over this. Going to put this on HOLD in the meantime.

**Comment by conorpendergrast** [comment]
Thanks Rory!

**Comment by conorpendergrast** [comment]
Still on hold, I see the PR is progressing

**Comment by puneetlath** [comment]
When we upgrade, it would be nice if we could upgrade to v0.19.6. We recently had a PR merged upstream into that version for [this bug](https://github.com/Expensify/App/issues/19220).

**Comment by conorpendergrast** [comment]
Still held on https://github.com/Expensify/App/pull/18507

**Comment by conorpendergrast** [comment]
Still held on #18507, which looks like it's getting closer

**Comment by conorpendergrast** [comment]
Still on hold

**Comment by alitoshmatov** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Upgrade react-native-web to 0.19 version
### What is the root cause of that problem?
We are currently using forked version of `react-native-web` and it `0.18` version with some patches.
### What changes do you think we should make in order to solve the problem?
We should switch back to using original `react-native-web` since all patches were also applied to upstream. We should install `react-native-web@0.19.7` which looks latest stable version.

We should also change webpack configs:
https://github.com/Expensify/App/blob/4e31da03133bd08b006370be4857a33487b4ba92/config/webpack/webpack.common.js#L186-L187
https://github.com/Expensify/App/blob/48ce5e4c70d1ced8d9a176bf203e07852b67b14b/.storybook/webpack.config.js#L27-L28

We should point `react-native` to `react-native-web` here and remove `react-native-web` pointing to `@expensify/react-native-web`
```diff
+ 'react-native$': 'react-native-web',
- 'react-native$': '@expensify/react-native-web',
- 'react-native-web': '@expensify/react-native-web',
```


### What alternative solutions did you explore? (Optional)

**Comment by getusha** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Upgrade react-native-web to version 0.19
### What is the root cause of that problem?
There are different issues which will be resolved by this upgrade, there is no specific root cause.
### What changes do you think we should make in order to solve the problem?

1. We can sync the latest version from upstream to our fork `@expensify/react-native-web`
2. We can just use `react-native-web` package directly inside E/App.

**There are breaking changes which should be changed/replace and removed.**
* `<Image>` `resizeMode` style is deprecated replace with `resizeMode` prop. 
https://github.com/Expensify/App/blob/4230f289341c8d6d9a8b780d54da0725d4764b2b/src/styles/styles.js#L3470
* `<TextInput>` instead of `keyboardType` prop use `inputMode`
example: https://github.com/Expensify/App/blob/4230f289341c8d6d9a8b780d54da0725d4764b2b/src/components/AmountTextInput.js#L48
* `<TextInput>` instead of `returnKeyType` prop use `enterKeyHint`
* `<TextInput>` instead of `editable` use `readOnly`
* `<TextInput>` instead of `numberOfLines` use `rows`
* `nativeId` is deprecated replace with `id`
* `focusable` is deprecated replaced with `tabIndex` we can make a component focusable by passing `tabIndex="0"` and not focusable by passing `tabIndex="-1"`
* `pointerEvents` prop is deprecated replace with `pointerEvents` style.
* ` accessibility* ` props are deprecated replace with `aria-*` and we replace `accessibilityRole` with `role`
* `Touchable*` components are deprecated we should replace it with `Pressable` i don't think we are using any `Touchable*` components for now.
* `selectable` prop is deprecated replace with `userSelect` style 
* `textAlignVertical` is deprecated replace with `verticalAlign` style
* `transform` array values are deprecated instead we should use spaced values e.g. **`transform: 'translateX(50%) translateY(50%)'`**
* `<Text>` **`position: relative`** is the default now, we can use position:'static' by adding a prop called `position` which has a value of position: 'static' by default to our custom `Text` component and we can change the position to relative/static when we have places which require relative positioning using the prop, if it is necessary.
### What alternative solutions did you explore? (Optional)
N/A

**Comment by alitoshmatov** [comment]
**Might be helpful**
There are some issues present when updating to `0.19` which is were not present in `0.18`(I checked by installing `react-native-web@0.18.0` 
I think `FlatList` is behaving differently, but couldn't pinpoint exact issue:
Here there is a space event there is one report action:
<img width="1440" alt="Screenshot 2023-08-05 at 13 47 48" src="https://github.com/Expensify/App/assets/59907218/027eb2be-3f32-48e7-ba27-148962f429b4">

Here hover state is lost when hovering over context menu, and leaving report action
https://github.com/Expensify/App/assets/59907218/f05135b8-55b3-4bb2-9fd8-d29d262e8f78

**Comment by mountiny** [comment]
Not on hold anymore! @janicduplessis coming from here https://github.com/Expensify/App/pull/19345 might have some ideas about the issues after updating to the new version.

**Comment by shubham1206agra** [comment]
Hey @mountiny 

Would you like to take a look at this PR https://github.com/Expensify/react-native-web/pull/22?

So that we can quickly proceed with the upgrade.

**Comment by roryabraham** [comment]
I think we should also strive to ditch our react-native-web fork as well. We can leverage patch-package as needed for custom code but generally should have upstream PRs for all custom code in the fork

**Comment by shubham1206agra** [comment]
> I think we should also strive to ditch our react-native-web fork as well. We can leverage patch-package as needed for custom code but generally should have upstream PRs for all custom code in the fork

Using upstream and patching may not be ideal here as there are 3 different places to patch. And all commits of the master come under v0.19 in RNW. So any change (which may change the line number) may break the patch and we have no control which changes will come upstream, which may cause problems.

**Comment by roryabraham** [comment]
> Using upstream and patching may not be ideal here as there are 3 different places to patch

@shubham1206agra we recently sponsored a new feature in patch-package which allows you to create multiple patch files for the same dependency and have them applied in sequence, so I think it makes it easier + more viable to use patch files when there are multiple different places to patch.

> all commits of the master come under v0.19 in RNW. So any change (which may change the line number) may break the patch and we have no control which changes will come upstream, which may cause problems

Sorry, I'm having trouble understanding what you mean by this.

**Comment by getusha** [comment]
@roryabraham Are there any plans to remove the fork `@expensify/react-native-web` if so, before or after this upgrade?

**Comment by roryabraham** [comment]
I kind of assume that we should be removing the `@expensify/react-native-web` fork for all the same reasons we're removing our React Native fork, and that we should make that change before this upgrade.

**Comment by shubham1206agra** [comment]
> I kind of assume that we should be removing the `@expensify/react-native-web` fork for all the same reasons we're removing our React Native fork, and that we should make that change before this upgrade.

Then we should start putting all changes upstream.

**Comment by roryabraham** [comment]
I just finished going through all the pull requests merged in our fork to try and understand all the custom code we have merged there and understand what we need to do to drop the fork and use the upstream instead. I tracked everything in [this spreadsheet](https://docs.google.com/spreadsheets/d/1PRqSzvxpPH5YRjRmLMWaokPTkFrBd4N4mSCnKtvjwHE/edit#gid=1594942597), and each PR essentially falls in one of three buckets.

### 🎉 Merged and released upstream 

- https://github.com/Expensify/react-native-web/pull/5
- https://github.com/Expensify/react-native-web/pull/9
- https://github.com/Expensify/react-native-web/pull/18

### ❌ Not merged upstream, but also not yet used in E/App so 😌 NAB for this upgrade

- https://github.com/Expensify/react-native-web/pull/11
- https://github.com/Expensify/react-native-web/pull/15
- https://github.com/Expensify/react-native-web/pull/20
- https://github.com/Expensify/react-native-web/pull/21

### 🩹 Needs a patch after this upgrade is complete

Additional context for all code related to VirtualizedLists: https://github.com/facebook/react-native/issues/35263

- https://github.com/Expensify/react-native-web/pull/2 – new upstream PR needed against @react-native/virtualized-lists subpackage of react-native, not react-native-web
- https://github.com/Expensify/react-native-web/pull/3 - same as https://github.com/Expensify/react-native-web/pull/2, should be folded into the same patch file and upstream PR
- https://github.com/Expensify/react-native-web/pull/7 – this _was_ merged upstream in React Native but has not been ported to upstream react-native-web yet
- https://github.com/Expensify/react-native-web/pull/16 – Need to create upstream PR

---

So with that, I feel more confident in moving forward with a PR in E/App to use the upstream dependency and add the 3 patch files we need.

**Comment by roryabraham** [comment]
Given my last comment, I am inclined to accept @getusha's proposal [here](https://github.com/Expensify/App/issues/16660#issuecomment-1666425053) because it has a good summary of upstream breaking changes.

@aimane-chnaif what do you think?

**Comment by tungmt** [comment]
@roryabraham Is there anything missing in my proposal? https://github.com/Expensify/App/issues/16660#issuecomment-1591366594

**Comment by roryabraham** [comment]
oh, apologies @tungmt for missing that. I think your earlier proposal is basically equivalent to @getusha's later proposal, so yours would get priority.

**Comment by getusha** [comment]
@roryabraham i included more missing changes that should be changed/replaced and it's quite different from @tungmt's earlier proposal.

**Comment by roryabraham** [comment]
Well, this is embarrassing. 😅  I apologize to everyone for moving to quickly / not carefully reviewing the proposals that we had here. My primary focus yesterday was to understand what we needed to do to get off our fork and onto the upstream.

@getusha's [proposal](https://github.com/Expensify/App/issues/16660#issuecomment-1666425053) builds on @tungmt's [proposal](https://github.com/Expensify/App/issues/16660#issuecomment-1591366594) by considering a few extra deprecations and changes:

- `<Image> resizeMode style is deprecated replace with resizeMode prop`
- `transform array values are deprecated instead we should use spaced values e.g. transform: 'translateX(50%) translateY(50%)'`

These are the only differences AFAICT, but I'm going to accept the more complete proposal.

**Comment by roryabraham** [comment]
@getusha there are a lot of things on HOLD for this issue so I'd love to see us complete it ASAP

**Comment by getusha** [comment]
>@getusha I am available to review during weekend if you can raise PR soon.

I will raise the PR as soon as possible. thank you!

**Comment by getusha** [comment]
Here is a draft PR https://github.com/Expensify/App/pull/24482
and i raised a PR for `react-native-key-command`  https://github.com/Expensify/react-native-key-command/pull/31 this will fix a conflict while installing packages.

@aimane-chnaif @roryabraham

**Comment by roryabraham** [comment]
Left some comments in the PR. Seems like it has stalled a bit, @getusha @aimane-chnaif any questions / concerns I can answer?

**Comment by roryabraham** [comment]
@getusha @aimane-chnaif @WoLewicki what's the latest news on the [RNW upgrade](https://github.com/Expensify/App/pull/24482)?

**Comment by getusha** [comment]
>@getusha @aimane-chnaif @WoLewicki what's the latest news on the https://github.com/Expensify/App/pull/24482?

@roryabraham as @aimane-chnaif mentioned we are investigating [this](https://github.com/Expensify/App/pull/24482#issuecomment-1705049731) issue which is already fixed but needed more explanation about the FreezeWrapper.

**Comment by conorpendergrast** [comment]
@kevinksullivan I'm on parental leave; re-assigning! This doesn't seem like to needs your work until payment

**Comment by arthurmfgtab** [comment]
Hey, I'm Arthur from Callstack and I'll start working on this 🙋

**Comment by iwiznia** [comment]
What's the state of this?

**Comment by getusha** [comment]
@iwiznia PR is already deployed to prod, it is awaiting a payment
cc @roryabraham @mountiny @kevinksullivan

**Comment by getusha** [comment]
@iwiznia Yeah, we are now using `react-native-web` 0.19+ version https://github.com/Expensify/App/blob/main/package.json#L161 and `inputMode` is supported.

**Comment by roryabraham** [comment]
Updating the bounty to $3500 as this was a larger issue with many sub-issues and conflicts

**Comment by kevinksullivan** [comment]
Payments made
